node/src/accountant/scanners/pending_payable_scanner/utils.rs

```diff
@@ -1,191 +1,1184 @@
+            (None, _) => None,
+            (Some(retry), _) => Some(retry),
+        }
+    }
+
+    pub(super) fn register_confirmed_tx(&mut self, confirmation: NormalTxConfirmation) {
+        self.confirmations.normal_confirmations.push(confirmation);
+    }
+
+    pub(super) fn register_failure_reclaim(&mut self, reclaim: TxReclaim) {
+        self.confirmations.reclaims.push(reclaim)
+    }
+
+    pub(super) fn register_new_failure(&mut self, failed_tx: FailedTx) {
+        self.failures
+            .tx_failures
+            .push(PresortedTxFailure::NewEntry(failed_tx));
+    }
+
+    pub(super) fn register_finalization_of_unproven_failure(&mut self, tx_hash: TxHash) {
+        self.failures
+            .tx_failures
+            .push(PresortedTxFailure::RecheckCompleted(tx_hash));
+    }
+
+    pub(super) fn register_rpc_failure(&mut self, status_update: FailedValidationByTable) {
+        self.failures.tx_receipt_rpc_failures.push(status_update);
+    }
+}
+
+#[derive(Debug, Default, PartialEq, Eq, Clone)]
+pub struct DetectedConfirmations {
+    pub normal_confirmations: Vec<NormalTxConfirmation>,
+    pub reclaims: Vec<TxReclaim>,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct NormalTxConfirmation {
+    pub tx: SentTx,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct TxReclaim {
+    pub reclaimed: SentTx,
+}
+
+impl DetectedConfirmations {
+    pub(super) fn is_empty(&self) -> bool {
+        self.normal_confirmations.is_empty() && self.reclaims.is_empty()
+    }
+}
+
+#[derive(Debug, Default, PartialEq, Eq, Clone)]
+pub struct DetectedFailures {
+    pub tx_failures: Vec<PresortedTxFailure>,
+    pub tx_receipt_rpc_failures: Vec<FailedValidationByTable>,
+}
+
+impl DetectedFailures {
+    fn requires_retry(&self) -> Option<Retry> {
+        if self.tx_failures.is_empty() && self.tx_receipt_rpc_failures.is_empty() {
+            None
+        } else if !self.tx_failures.is_empty() {
+            Some(Retry::RetryPayments)
+        } else {
+            Some(Retry::RetryTxStatusCheckOnly)
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum PresortedTxFailure {
+    NewEntry(FailedTx),
+    RecheckCompleted(TxHash),
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum FailedValidationByTable {
+    SentPayable(FailedValidation<TxStatus>),
+    FailedPayable(FailedValidation<FailureStatus>),
+}
+
+impl From<(TxReceiptError, TxStatus)> for FailedValidationByTable {
+    fn from((tx_receipt_error, current_status): (TxReceiptError, TxStatus)) -> Self {
+        match tx_receipt_error.tx_hash {
+            TxHashByTable::SentPayable(tx_hash) => Self::SentPayable(FailedValidation::new(
+                tx_hash,
+                BlockchainErrorKind::AppRpc(tx_receipt_error.err.into()),
+                current_status,
+            )),
+
+            TxHashByTable::FailedPayable(tx_hash) => {
+                unreachable!(
```

Comment:
Using `unreachable!()` adds uncertainity in the production code and opens a risk of panic. I know that we are under the assumption that the code will never reach this place, but still, it delegates that responsibility to the developer using this.

TL;DR - If we can escape using the `unreachable!()` please do that. If I were you, I would either use `try_from`, or build it using a constructor.